### ISSUE TITLE
Wrong inclusion of "location" metadata in geo:json notifications

### ISSUE THREAD

**Issue by fgalan** [telefonicaid/fiware-orion#3045]
Entity created like this:

```
curl -vvvv localhost:1026/v2/entities -s -S --header 'Content-Type: application/json' -d @- <<EOF
{
  "id": "E",
  "type": "T",
  "name":{
    "value": "SD3",
    "type": "Text"
  },
  "perimeter": {
    "type":"geo:json",
    "value":{
    "type": "Polygon",
    "coordinates": [
                    [
                        [-13.806454999999991,33.46727800000001,0],
                        [-13.806707999999981,33.46723400000001,0],
                        [-13.806863999999999,33.46720599999997,0],
                        [-13.807017000000008,33.46719100000001,0],
                        [-13.806454999999991,33.46727800000001,0]
                    ]
                ]
        }
   }
}
EOF
```

A subscription on that entity is created like this:

```
curl -vvvvv localhost:1026/v2/subscriptions -s -S --header 'Content-Type: application/json' -d @- <<EOF
{
  "description": "A subscription to get info about E",
  "subject": {
    "entities": [
      {
        "id": "E",
        "type": "T"
      }
    ],
    "condition": {
      "attrs": [
      ]
    }
  },
  "notification": {
    "http": {
      "url": "http://localhost:1028/notify"
    },
    "attrs": [     
    ]
  } 
}
EOF
```

Notifications are triggered with something like this:

```
curl -vvvv localhost:1026/v2/entities/E/attrs/name -X PUT -s -S --header 'Content-Type: application/json' -d @- <<EOF
{
  "value": "$1"
}
EOF
```

Notification wrongly include a `location` metadata with `"WGS84"` value for the `perimeter` atribute, i.e. mixing with the old NGSIv1-base mechanism. Note that GET on the same entity is not including it. Thus, GET is ok but notifications isn't.

Some additional remarks:

* Tested with geo:json but it may happen with geo:point, geo:line, geo:box and geo:polygon. All them should be covered by the associated .test.
* The problem seems to happen both in NGSIv2 and legacy NGSIv1 notifications. The associated .test should cover both.

**Comment by arigliano** [comment]
Hello, I would like to work on this issue. Can you assign it to me?

**Comment by fgalan** [comment]
Of course. Done!

**Comment by arigliano** [comment]
After PR #3122 clarifications and related logic fixes, looking deeper this issue, the problem seems to be in the following, the entitiesQuery calls does not receive the ApiVersion:
```
 static bool processOnChangeConditionForSubscription
(
  const EntityIdVector&            enV,
  const StringList&                attrL,
  const std::vector<std::string>&  metadataV,
  ConditionValueList*              condValues,
  const std::string&               subId,
  const HttpInfo&                  notifyHttpInfo,
  RenderFormat                     renderFormat,
  const std::string&               tenant,
  const std::string&               xauthToken,
  const std::vector<std::string>&  servicePathV,
  const Restriction*               resP,
  const std::string&               fiwareCorrelator,
  const std::vector<std::string>&  attrsOrder,
  bool                             blacklist
)
{
  std::string                   err;
  NotifyContextRequest          ncr;
  ContextElementResponseVector  rawCerV;
  StringList                    emptyList;
  StringList                    metadataList;

  metadataList.fill(metadataV);
  if (!blacklist && !entitiesQuery(enV, attrL, *resP, &rawCerV, &err, true, tenant, servicePathV))
  {
    ncr.contextElementResponseVector.release();
    rawCerV.release();

    return false;
  }
  else if (blacklist && !entitiesQuery(enV, metadataList, *resP, &rawCerV, &err, true, tenant, servicePathV))
  {
    ncr.contextElementResponseVector.release();
    rawCerV.release();

    return false;
  }
```
Then in the **entitiesQuery** function, the API version is not present in the call, and it seems to take 1 as default:

```
 // Build CER from BSON retrieved from DB
    docs++;
    LM_T(LmtMongo, ("retrieved document [%d]: '%s'", docs, r.toString().c_str()));
    ContextElementResponse*  cer = new ContextElementResponse(r, attrL, includeEmpty, apiVersion);

    // Add builtin attributes and metadata (only in NGSIv2)
    if (apiVersion == V2)
    {
      addBuiltins(cer);
    }
```
Finally, the ContextElementResponse constructor, receives "1" as apiVersion (even if the actual request was 2, as the apiVersion was not passed in the previous calls (processOnChangeConditionForSubscription -> (several) entitiesQuery -> new ContextElementResponse  ) : 

In the #68020b7 commit, done in the PR, the constructor has the dedicated part to correctly handle location metadata:
 

```
        /* The flag below indicates that a location metadata was found during iteration.
        * It needs to the NGSIV1 check below, in order to add it if the flag is false */
        if (md->name == NGSI_MD_LOCATION)
        {
          noLocationMetadata = false;
        }
      }
    }
     if (apiVersion == V1)
    {
      /* Setting location metadata (if location attr found
       *  and the location metadata was not present) */
      if ((locAttr == ca.name) && (ca.type != GEO_POINT) && noLocationMetadata)
      {
        /* Note that if attribute type is geo:point then the user is using the "new way"
         * of locating entities in NGSIv1, thus location metadata is not rendered */
        Metadata* md = new Metadata(NGSI_MD_LOCATION, "string", LOCATION_WGS84);
        caP->metadataVector.push_back(md);
      }
    }
```
But it receives wrongly V1. The solution could be to pass in processOnChangeConditionForSubscription the apiVersion, in order to pass it in the entitiesQuery calls in the body. An alternative could be to the renderFormat as apiVersion to be passed in the entitiesQuery calls.

**Comment by fgalan** [comment]
> The solution could be to pass in processOnChangeConditionForSubscription the apiVersion, in order to pass it in the entitiesQuery calls in the body. An alternative could be to the renderFormat as apiVersion to be passed in the entitiesQuery calls.

Without having analyzed the issue so deep as you have done (thanks! :) I think best alterantive would be:

> The solution could be to pass in processOnChangeConditionForSubscription the apiVersion, in order to pass it in the entitiesQuery calls in the body

In fact, I think the code was that way in the past but we removed it as we didn't need (until now)

With regards to

> An alternative could be to the renderFormat as apiVersion to be passed in the entitiesQuery calls.

I think it is not a good idea, as it would involve overload the RenderFormat semantics with an aspect not related with it. RenderFormat is used only by notifications and I understand (tell me if I'm wrong) the issue affect both notifications and GET responses.

**Comment by fgalan** [comment]
Fixed by PR #3346
